File: electron/db/repositories/PrivacyRuleRepository.ts
/**
 * PrivacyRuleRepository：隐私规则数据访问层 + 规则匹配引擎
 * 匹配动作由规则类型决定：
 *   - app_name / process_name 命中 → skip（完全跳过，不截图不 OCR）
 *   - window_title / url 命中 → placeholder（占位符，记录隐私窗口被保护）
 *   - 未命中 → allow
 */
import { randomUUID } from 'node:crypto'
import type { PrivacyRule, PrivacyRuleType, PrivacyMatchMode, PrivacyMatchResult } from '@/types'
import { getDatabase } from '../database'

interface PrivacyRuleRow {
  id: string
  type: string
  pattern: string
  match_mode: string
  enabled: number
}

function rowToRule(row: PrivacyRuleRow): PrivacyRule {
  return {
    id: row.id,
    type: row.type as PrivacyRuleType,
    pattern: row.pattern,
    matchMode: row.match_mode as PrivacyMatchMode,
    enabled: row.enabled === 1
  }
}

interface RuleParams {
  id: string
  type: string
  pattern: string
  match_mode: string
  enabled: number
}

function ruleToParams(rule: PrivacyRule): RuleParams {
  return {
    id: rule.id,
    type: rule.type,
    pattern: rule.pattern,
    match_mode: rule.matchMode,
    enabled: rule.enabled ? 1 : 0
  }
}

/** 单条规则对单个目标值的匹配判定 */
function matchValue(value: string, pattern: string, mode: PrivacyMatchMode): boolean {
  if (!pattern) return false
  const v = value || ''
  switch (mode) {
    case 'contains':
      return v.toLowerCase().includes(pattern.toLowerCase())
    case 'equals':
      return v.toLowerCase() === pattern.toLowerCase()
    case 'regex': {
      try {
        return new RegExp(pattern, 'i').test(v)
      } catch {
        return false
      }
    }
    default:
      return false
  }
}

/** 根据规则类型决定过滤动作 */
function actionForType(type: PrivacyRuleType): 'skip' | 'placeholder' {
  return type === 'app_name' || type === 'process_name' ? 'skip' : 'placeholder'
}

export const PrivacyRuleRepository = {
  insert(rule: Omit<PrivacyRule, 'id'>): PrivacyRule {
    const db = getDatabase()
    const id = randomUUID()
    const params = ruleToParams({ ...rule, id })
    db.prepare(
      `INSERT INTO privacy_rules (id, type, pattern, match_mode, enabled)
       VALUES (@id, @type, @pattern, @match_mode, @enabled)`
    ).run(params)
    const created = this.getAll().find(r => r.id === id)
    if (!created) throw new Error(`PrivacyRule insert failed for id=${id}`)
    return created
  },

  update(id: string, patch: Partial<PrivacyRule>): PrivacyRule | null {
    const existing = this.getAll().find(r => r.id === id)
    if (!existing) return null
    const merged: PrivacyRule = { ...existing, ...patch, id }
    const params = ruleToParams(merged)
    const db = getDatabase()
    db.prepare(
      `UPDATE privacy_rules SET type = @type, pattern = @pattern, match_mode = @match_mode, enabled = @enabled WHERE id = @id`
    ).run(params)
    return this.getAll().find(r => r.id === id) ?? null
  },

  delete(id: string): boolean {
    const db = getDatabase()
    const result = db.prepare('DELETE FROM privacy_rules WHERE id = ?').run(id)
    return result.changes > 0
  },

  getAll(): PrivacyRule[] {
    const db = getDatabase()
    const rows = db
      .prepare('SELECT * FROM privacy_rules ORDER BY rowid ASC')
      .all() as PrivacyRuleRow[]
    return rows.map(rowToRule)
  },

  getEnabled(): PrivacyRule[] {
    const db = getDatabase()
    const rows = db
      .prepare('SELECT * FROM privacy_rules WHERE enabled = 1 ORDER BY rowid ASC')
      .all() as PrivacyRuleRow[]
    return rows.map(rowToRule)
  },

  /**
   * 规则匹配引擎：依次用启用的规则匹配传入的应用/进程/窗口标题/URL。
   * 优先级：app_name/process_name（skip）先于 window_title/url（placeholder）。
   * 返回首个命中规则对应的动作；未命中返回 allow。
   */
  matchRule(
    appName: string,
    processName: string,
    windowTitle: string,
    url: string
  ): PrivacyMatchResult {
    const rules = this.getEnabled()
    // 第一轮：检查 skip 类规则（app_name / process_name）
    for (const rule of rules) {
      if (rule.type === 'app_name') {
        if (matchValue(appName, rule.pattern, rule.matchMode)) {
          return { action: 'skip', matchedRule: rule }
        }
      } else if (rule.type === 'process_name') {
        if (matchValue(processName, rule.pattern, rule.matchMode)) {
          return { action: 'skip', matchedRule: rule }
        }
      }
    }
    // 第二轮：检查 placeholder 类规则（window_title / url）
    for (const rule of rules) {
      if (rule.type === 'window_title') {
        if (matchValue(windowTitle, rule.pattern, rule.matchMode)) {
          return { action: 'placeholder', matchedRule: rule }
        }
      } else if (rule.type === 'url') {
        if (matchValue(url, rule.pattern, rule.matchMode)) {
          return { action: 'placeholder', matchedRule: rule }
        }
      }
    }
    return { action: 'allow', matchedRule: null }
  }
}

export { matchValue, actionForType }

File: src-tauri/src/repositories/privacy_rule_repository.rs
//! PrivacyRuleRepository：隐私规则数据访问层 + 规则匹配引擎（对应 electron/db/repositories/PrivacyRuleRepository.ts）
//!
//! 匹配动作由规则类型决定：
//!   - app_name / process_name 命中 → skip（完全跳过，不截图不 OCR）
//!   - window_title / url 命中 → placeholder（占位符，记录隐私窗口被保护）
//!   - 未命中 → allow

use rusqlite::{params, Connection};

use crate::db::database::get_database;
use crate::models::{
    PrivacyAction, PrivacyMatchMode, PrivacyMatchResult, PrivacyRule, PrivacyRuleType,
};

fn row_to_rule(row: &rusqlite::Row<'_>) -> rusqlite::Result<PrivacyRule> {
    let type_str: String = row.get("type")?;
    let match_mode_str: String = row.get("match_mode")?;
    Ok(PrivacyRule {
        id: row.get("id")?,
        rule_type: PrivacyRuleType::from_str(&type_str),
        pattern: row.get("pattern")?,
        match_mode: PrivacyMatchMode::from_str(&match_mode_str),
        enabled: row.get::<_, i64>("enabled")? != 0,
    })
}

/// 单条规则对单个目标值的匹配判定
pub fn match_value(value: &str, pattern: &str, mode: PrivacyMatchMode) -> bool {
    if pattern.is_empty() {
        return false;
    }
    match mode {
        PrivacyMatchMode::Contains => value.to_lowercase().contains(&pattern.to_lowercase()),
        PrivacyMatchMode::Equals => value.to_lowercase() == pattern.to_lowercase(),
        PrivacyMatchMode::Regex => {
            let re = match regex::Regex::new(&format!("(?i){}", pattern)) {
                Ok(r) => r,
                Err(_) => return false,
            };
            re.is_match(value)
        }
    }
}

/// 根据规则类型决定过滤动作
pub fn action_for_type(rule_type: PrivacyRuleType) -> PrivacyAction {
    match rule_type {
        PrivacyRuleType::AppName | PrivacyRuleType::ProcessName => PrivacyAction::Skip,
        _ => PrivacyAction::Placeholder,
    }
}

pub struct PrivacyRuleRepository;

impl PrivacyRuleRepository {
    pub fn insert(rule: PrivacyRule) -> anyhow::Result<PrivacyRule> {
        let id = if rule.id.is_empty() {
            uuid::Uuid::new_v4().to_string()
        } else {
            rule.id.clone()
        };
        let mut rule = rule;
        rule.id = id.clone();

        let conn = get_database()?;
        conn.execute(
            "INSERT INTO privacy_rules (id, type, pattern, match_mode, enabled)
             VALUES (?1, ?2, ?3, ?4, ?5)",
            params![
                rule.id,
                rule.rule_type.as_str(),
                rule.pattern,
                rule.match_mode.as_str(),
                rule.enabled as i64,
            ],
        )?;
        drop(conn);
        Self::get_all()?
            .into_iter()
            .find(|r| r.id == id)
            .ok_or_else(|| anyhow::anyhow!("PrivacyRule insert failed for id={}", id))
    }

    pub fn update(id: &str, patch: PrivacyRule) -> anyhow::Result<Option<PrivacyRule>> {
        let existing = match Self::get_all()?.into_iter().find(|r| r.id == id) {
            Some(e) => e,
            None => return Ok(None),
        };
        let merged = merge_rule(existing, patch, id);
        let conn = get_database()?;
        conn.execute(
            "UPDATE privacy_rules SET type = ?1, pattern = ?2, match_mode = ?3, enabled = ?4 WHERE id = ?5",
            params![
                merged.rule_type.as_str(),
                merged.pattern,
                merged.match_mode.as_str(),
                merged.enabled as i64,
                merged.id,
            ],
        )?;
        drop(conn);
        Ok(Self::get_all()?.into_iter().find(|r| r.id == id))
    }

    pub fn delete(id: &str) -> anyhow::Result<bool> {
        let conn = get_database()?;
        let changes = conn.execute("DELETE FROM privacy_rules WHERE id = ?1", params![id])?;
        Ok(changes > 0)
    }

    pub fn get_all() -> anyhow::Result<Vec<PrivacyRule>> {
        let conn = get_database()?;
        let mut stmt = conn.prepare("SELECT * FROM privacy_rules ORDER BY rowid ASC")?;
        let rules = stmt
            .query_map([], row_to_rule)?
            .collect::<rusqlite::Result<Vec<_>>>()?;
        Ok(rules)
    }

    pub fn get_enabled() -> anyhow::Result<Vec<PrivacyRule>> {
        let conn = get_database()?;
        let mut stmt =
            conn.prepare("SELECT * FROM privacy_rules WHERE enabled = 1 ORDER BY rowid ASC")?;
        let rules = stmt
            .query_map([], row_to_rule)?
            .collect::<rusqlite::Result<Vec<_>>>()?;
        Ok(rules)
    }

    /// 规则匹配引擎
    pub fn match_rule(
        app_name: &str,
        process_name: &str,
        window_title: &str,
        url: &str,
    ) -> anyhow::Result<PrivacyMatchResult> {
        let rules = Self::get_enabled()?;
        // 第一轮：检查 skip 类规则（app_name / process_name）
        for rule in &rules {
            match rule.rule_type {
                PrivacyRuleType::AppName => {
                    if match_value(app_name, &rule.pattern, rule.match_mode) {
                        return Ok(PrivacyMatchResult {
                            action: PrivacyAction::Skip,
                            matched_rule: Some(rule.clone()),
                        });
                    }
                }
                PrivacyRuleType::ProcessName => {
                    if match_value(process_name, &rule.pattern, rule.match_mode) {
                        return Ok(PrivacyMatchResult {
                            action: PrivacyAction::Skip,
                            matched_rule: Some(rule.clone()),
                        });
                    }
                }
                _ => {}
            }
        }
        // 第二轮：检查 placeholder 类规则（window_title / url）
        for rule in &rules {
            match rule.rule_type {
                PrivacyRuleType::WindowTitle => {
                    if match_value(window_title, &rule.pattern, rule.match_mode) {
                        return Ok(PrivacyMatchResult {
                            action: PrivacyAction::Placeholder,
                            matched_rule: Some(rule.clone()),
                        });
                    }
                }
                PrivacyRuleType::Url => {
                    if match_value(url, &rule.pattern, rule.match_mode) {
                        return Ok(PrivacyMatchResult {
                            action: PrivacyAction::Placeholder,
                            matched_rule: Some(rule.clone()),
                        });
                    }
                }
                _ => {}
            }
        }
        Ok(PrivacyMatchResult {
            action: PrivacyAction::Allow,
            matched_rule: None,
        })
    }

    /// 播种默认隐私规则（对应 PrivacyGuard.seedDefaultRules）
    pub fn seed_default_rules() -> anyhow::Result<()> {
        let existing = Self::get_all()?;
        if !existing.is_empty() {
            return Ok(());
        }
        let defaults = vec![
            // 进程级 skip
            PrivacyRule {
                id: String::new(),
                rule_type: PrivacyRuleType::ProcessName,
                pattern: "1password".to_string(),
                match_mode: PrivacyMatchMode::Contains,
                enabled: true,
            },
            PrivacyRule {
                id: String::new(),
                rule_type: PrivacyRuleType::ProcessName,
                pattern: "keepass".to_string(),
                match_mode: PrivacyMatchMode::Contains,
                enabled: true,
            },
            PrivacyRule {
                id: String::new(),
                rule_type: PrivacyRuleType::ProcessName,
                pattern: "bitwarden".to_string(),
                match_mode: PrivacyMatchMode::Contains,
                enabled: true,
            },
            // 银行/支付类窗口标题 placeholder
            PrivacyRule {
                id: String::new(),
                rule_type: PrivacyRuleType::WindowTitle,
                pattern: "银行".to_string(),
                match_mode: PrivacyMatchMode::Contains,
                enabled: true,
            },
            PrivacyRule {
                id: String::new(),
                rule_type: PrivacyRuleType::WindowTitle,
                pattern: "支付宝".to_string(),
                match_mode: PrivacyMatchMode::Contains,
                enabled: true,
            },
            PrivacyRule {
                id: String::new(),
                rule_type: PrivacyRuleType::WindowTitle,
                pattern: "微信支付".to_string(),
                match_mode: PrivacyMatchMode::Contains,
                enabled: true,
            },
        ];
        for rule in defaults {
            Self::insert(rule)?;
        }
        Ok(())
    }
}

fn merge_rule(mut existing: PrivacyRule, patch: PrivacyRule, id: &str) -> PrivacyRule {
    existing.rule_type = patch.rule_type;
    if !patch.pattern.is_empty() {
        existing.pattern = patch.pattern;
    }
    existing.match_mode = patch.match_mode;
    existing.enabled = patch.enabled;
    existing.id = id.to_string();
    existing
}
